Sessions: add charge duration

naltatis opened the pull request:
fixes #8837

- Add charge duration to session (UI & CSV)
- Refactored and tested duration formatting

![Bildschirmfoto 2023-08-05 um 18 47 30](https://github.com/evcc-io/evcc/assets/152287/f0ce6076-fb02-4d30-a250-5fa31194797f)

![Bildschirmfoto 2023-08-05 um 18 46 22](https://github.com/evcc-io/evcc/assets/152287/73209268-a082-423b-8cac-b1f5306c3615)

Todos
- [x] show in session details
- [x] improve md breakpoint (use configurable columns as mobile)
- [x] integration tests
- [x] calculate avg power per session based on energy and duration (ui only).

naltatis committed:
Sessions: add charge duration

naltatis committed:
use small non-breaking space

StefanSchoof commented:
Wenn ich mir den Screenshot anschauen, frage ich mich, ob Dauer die Dauer des aktiven Ladens oder die Dauer des angeschlossen sein ist. (Die CSV Überschrift verrät es) Kann man irgendwo noch eine Erläuterung dazu unterbringen?

naltatis commented:
Wollte den Titel kurz halten weil es sonst die Spalte so breit drückt. Ladedauer wäre aber vermutlich auch ok.

StefanSchoof commented:
Ja, verstehe den Konflikt und auch das man die Frage wahrscheinlich nur hat, wenn man die Ansicht noch nicht kennt.

naltatis committed:
configurable columns if needed

naltatis commented:
sm
![Bildschirmfoto 2023-08-06 um 00 46 46](https://github.com/evcc-io/evcc/assets/152287/41f6e612-2edc-4055-89bf-0cac68ac2b40)

md
![Bildschirmfoto 2023-08-06 um 00 44 50](https://github.com/evcc-io/evcc/assets/152287/0ab0f70a-cdf3-4f6c-a198-a4bc7430be41)

xl
![Bildschirmfoto 2023-08-06 um 00 46 56](https://github.com/evcc-io/evcc/assets/152287/2758cc31-87dd-469b-b0e4-991f246b17f4)

premultiply commented:
Es sollte klar sein, dass dies die summierte Nettoladezeit ist: chargeDuration.
Sprich die Summe der Dauer der einzelnen Zeitabschnitte während der aktuellen Session in dem der Ladepunkt im Zustand "laden" ist, also Netzspannung am Fahrzeug anliegt bzw. das Schütz des Ladepunkts angezogen ist.

naltatis committed:
added avg power, added to session modal, fixed tests

naltatis committed:
toml

andig committed:
Versicharge: update register map (#9287)

panzerdev committed:
VW ID: implement api.Resurrector to start charging from sleep (#9276)

andig committed:
chore: minor

naltatis committed:
Settings/Savings: persist unsaved values regularly (#9257)

andig committed:
chore: improve trimming string from bytes

andig committed:
chore: upgrade and fork go-powerwall

andig committed:
chore: fix forked go-powerwall

andig committed:
chore: lazy-start heartbeat

andig commented:
Was ist hier die Ladedauer: die Nettoladezeit oder die zwischen Start und Ende? Falls Netto müssten wir nochmal schauen, ob wir die da wo der Charger sie nicht bereit stellt auch korrekt ausrechnen.

andig commented on i18n/de.toml:
```diff
@@ -145,7 +145,7 @@ selfConsumption = "Eigenverbrauch"
 charged = "Geladen"
 co2 = "⌀ CO₂"
-duration = "Dauer"
+duration = "Ladezeit"
```
Korrekt wäre Ladedauer. Es ist ja keine (Uhr)Zeit.

andig replied:
Hilft es (im Session Log) hier den "Lade" Präfix zu ergänzen? Was für eine Dauer könnte es noch sein?

naltatis replied:
Siehe Diskussion oben https://github.com/evcc-io/evcc/pull/9263#issuecomment-1666563645

> Was für eine Dauer könnte es noch sein?

Der Zeitraum vom Anstecke bis zum Abstecken. Also inkl. der Zeit in der nicht geladen wird.

Ich habs jetzt auf "Ladedauer" geändert. Im normalen Sprachgebrauch wird das sowieso Synonym verwendet und ich sehe hier keine Verwechselungsgefahr mit einem Zeitpunkt.

andig commented on i18n/de.toml:
```diff
@@ -285,7 +285,9 @@ started = "Startzeit"
 [sessions]
+avgPower = "⌀ Leistung"
 avgPrice = "⌀ Preis"
+chargeDuration = "Ladezeit"
```
Dito

andig commented on core/db/session.go:
```diff
@@ -30,6 +30,7 @@ type Session struct {
 	MeterStart      *float64  `json:"meterStart" csv:"Meter Start (kWh)" gorm:"column:meter_start_kwh"`
 	MeterStop       *float64  `json:"meterStop" csv:"Meter Stop (kWh)" gorm:"column:meter_end_kwh"`
 	ChargedEnergy   float64   `json:"chargedEnergy" csv:"Charged Energy (kWh)" gorm:"column:charged_kwh"`
+	ChargeDuration  *float64  `json:"chargeDuration" csv:"Charge Duration (s)" gorm:"column:charge_duration_s"`
```
Warum keine `time.Duration` einschl. Formatierung in `(t *Sessions) writeRow`?

naltatis replied:
Das führt dazu, dass wir den Wert in `ns` in der Datenbank abspeichern und auch im JSON so bekommen. Ich find das etwas unhandlich. Die Änderung in https://github.com/evcc-io/evcc/pull/9309 macht zwar, dass das im CSV dann menschenlesbar formatiert ist (12m22s), will man die Daten aber weiterverarbeiten muss man wieder zurückparsen. Numbers kann das scheinbar, aber ich hab das Gefühl die Änderung machts eher komplizierter als einfacher.

andig replied:
Die Verarbeitung von time.Duration ist ab Werk enthalten. Sekunden hätten nur einen Vorteil wenn ich die direkt aus der DB verarbeiten wollte?

andig commented on core/loadpoint_session.go:
```diff
@@ -73,6 +73,8 @@ func (lp *Loadpoint) stopSession() {
 	s.PricePerKWh = lp.sessionEnergy.PricePerKWh()
 	s.Co2PerKWh = lp.sessionEnergy.Co2PerKWh()
 	s.ChargedEnergy = lp.sessionEnergy.TotalWh() / 1e3
+	chargedSeconds := lp.chargeDuration.Seconds()
```
Falls `time.Duration` dann vllt. sogar `.Round(time.Minute)` runden?

naltatis replied:
s.o. ich verstehe den Charme hier im Go Typesystem zu bleiben, aber `ns` in der DB zu halten fühlt sich wie Overkill an. Vielleicht übersehe ich hier auch was, hab aber auf die schnelle keine Konfigurationsmöglichkeiten gefunden.

andig commented:
Siehe auch https://github.com/evcc-io/evcc/pull/9309 für Export. Macht nur den Code "schöner", Duration sollte auch jetzt schon als Go Format (0:00s) funktionieren.

andig committed:
Fix potential busy loop with short timeouts (#9320)

andig committed:
Heidelberg EC: handle failsafe timeout (#9322)

andig committed:
chore: fix conversion

poohnet committed:
Fix & simplify SMA Data Manager template (#9323)

andig committed:
Plugins: let charger support api.ChargerEx (#9331)

andig committed:
chore: use charge status mapper (#9305)

premultiply committed:
Docs: add missing 1p3p capabilities (#9334)

naltatis committed:
Fix odometer number format (#9345)

premultiply committed:
Enphase: fix soc and power (#9332)

* fix soc and power readings

* remove unspecified grid import energy

* remove cache

* wip

premultiply committed:
Add Acrel ADW300 (#9330)

VolkerK62 committed:
Shelly: set meter default channel 0 (#9362)

DerAndereAndi committed:
Improve EEBUS isCharging detection (#9354)

- Make sure the check using an external meter is identical to using an internal meter
- Fix check using internal meter to compare sum of all charge currents to idlefactor * min. single phase charge current

andig committed:
Ensure currents get synced (#9369)

andig committed:
Warp2: refactor ChargerEx implementation

andig committed:
Make csv export more flexible (#9309)

andig commented:
ping @naltatis

naltatis committed:
Merge branch 'master' into feature/store_charge_duration

naltatis committed:
fix: Dauer

naltatis committed:
fix: Dauer

naltatis committed:
fix: Dauer

naltatis committed:
time.Duration in db, s > ns formatting in ui

naltatis committed:
fix tests

naltatis commented:
@andig wenn du nichts mehr hast würd ich's rein nehmen
